### PR TITLE
[v2] RISC-V: Support device tree

### PR DESCRIPTION
**Pull request by gagachang**
This PR is version 2 of https://github.com/OP-TEE/optee_os/pull/6147.
Since some existed device tree functions in ARM can also be used by other architectures,
we create `core/kernel/boot.c` to move those functions into it.

Now RISC-V can get the external device tree provided by early boot stage.
And invoke aforementioned device tree functions to utilize the DTB.

### CODE REVIEW

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "core: arm: Move some DT functions to common kernel".
As this commit moves DT related functions, maybe it could go the in generic _kernel/dt.c_ instead of _kernel/boot.c_. That's just a n open question, not a strong requirement from me.

`dmqlkfjdsqlkfjksdmlkf`

**Review by etienne-lms — COMMENTED** [review]
Commit "core: kernel: Add API for getting external DT descriptor" must be squashed into  commit "core: arm: Move some DT functions to common kernel" unless what build is broken.

File: core/arch/riscv/kernel/entry.S [hunk 1]
```diff
@@ -188,6 +202,8 @@ UNWIND(	.cantunwind)
 	mv	s3, a0
 	STR	x0, THREAD_CORE_LOCAL_FLAGS(s3)
 
+	mv	a0, s1		/* s1 contains saved device tree address */
+	mv	a1, x0		/* unused */
```
Comment (etienne-lms) [inline core/arch/riscv/kernel/entry.S:206]: `x0` seems to hold some value. It should better be zeroed before being passed to `boot_init_primary_late()`. Unless it's fine as is as per _arch/riscv/kernel/boot.c_ implementation.

File: core/arch/riscv/kernel/boot.c [hunk 2]
```diff
@@ -31,6 +32,32 @@ unsigned long boot_args[4];
...
+}
+
+static void update_external_dt(void)
+{
+	struct dt_descriptor *dt = get_external_dt_desc();
+
+	if (!IS_ENABLED(CFG_EXTERNAL_DT))
+		return;
+
+	if (!dt->blob)
```
Comment (etienne-lms) [inline core/arch/riscv/kernel/boot.c:49]: test case `dt` is NULL?

**Comment by gagachang** [comment]
> As this commit moves DT related functions, maybe it could go the in generic _kernel/dt.c_ instead of _kernel/boot.c_. That's just a n open question, not a strong requirement from me.

`kernel/dt.c` is compiled only when `CFG_DT` is enabled.
But some functions such as `get_dt()` and `get_secure_dt()` should be independent with `CFG_DT`.
Maybe I can move dependent functions to `kernel/dt.c` and leave independent functions to `kernel/boot.c`

File: core/arch/riscv/kernel/entry.S (hunk 1 above)
Comment (gagachang) [reply core/arch/riscv/kernel/entry.S:206]: `x0` is hardwired to zero, same as ARM's `WZR/XZR`.

`mv a1, x0` is equivalent to `a1 = 0`.

File: core/arch/riscv/kernel/boot.c (hunk 2 above)
Comment (gagachang) [reply core/arch/riscv/kernel/boot.c:49]: That should be checked, thanks!

File: core/arch/riscv/kernel/entry.S (hunk 1 above)
Comment (etienne-lms) [reply core/arch/riscv/kernel/entry.S:206]: ok, understood. thanks.

File: core/arch/riscv/kernel/boot.c [hunk 3]
```diff
@@ -31,6 +33,32 @@ unsigned long boot_args[4];
...
+	return add_res_mem_dt_node(dt, "optee_core", CFG_TDDRAM_START,
+				   CFG_TDDRAM_SIZE);
+}
+
+static void update_external_dt(void)
+{
+	struct dt_descriptor *dt = get_external_dt_desc();
+
+	if (!IS_ENABLED(CFG_EXTERNAL_DT))
+		return;
```
Comment (etienne-lms) [inline core/arch/riscv/kernel/boot.c:48]: Can remove these 2 lines. When `CFG_EXTERNAL_DT` is disabled, `dt` is NULL.

File: core/arch/riscv/kernel/boot.c [hunk 4]
```diff
@@ -110,6 +138,9 @@ void boot_init_primary_early(unsigned long pageable_part __unused,
 void boot_init_primary_late(unsigned long fdt __unused,
 			    unsigned long tos_fw_config __unused)
 {
+	init_external_dt(fdt);
```
Comment (etienne-lms) [inline core/arch/riscv/kernel/boot.c:141]: remove `__unused` at line 138.

File: core/include/kernel/boot.h [hunk 5]
```diff
@@ -92,6 +104,43 @@ int boot_core_release(size_t core_idx, paddr_t entry);
...
+ * add_res_mem_dt_node() - Create "reserved-memory" parent and child nodes.
+ * @dt:		Pointer to a device tree descriptor which has DTB.
+ * @name:	Name of the child node.
+ * @pa:		Physical address of specific reserved memory region.
+ * @size:	Size of specific reserved memory region.
+ *
+ * Returns 0 if successes, otherwise -1.
+ *
+ * 1. Create /reserved-memory DTB node if it is not found in DTB,
+ * 2. Add /reserved-memory/ child node which has given name and memory region.
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/include/kernel/boot.h]: Removed. Thanks

File: core/include/kernel/boot.h [hunk 6]
```diff
@@ -92,6 +104,43 @@ int boot_core_release(size_t core_idx, paddr_t entry);
...
+			const char *subnode);
+
+/*
+ * add_res_mem_dt_node() - Create "reserved-memory" parent and child nodes.
+ * @dt:		Pointer to a device tree descriptor which has DTB.
+ * @name:	Name of the child node.
+ * @pa:		Physical address of specific reserved memory region.
+ * @size:	Size of specific reserved memory region.
+ *
+ * Returns 0 if successes, otherwise -1.
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/include/kernel/boot.h]: Done

File: core/include/kernel/boot.h [hunk 7]
```diff
@@ -92,6 +104,43 @@ int boot_core_release(size_t core_idx, paddr_t entry);
...
+ */
+void init_external_dt(unsigned long phys_dt);
+
+/*
+ * add_dt_path_subnode() - Add new child node into a parent node.
+ * @dt:		Pointer to a device tree descriptor which has DTB.
+ * @path:	Path to the parent node.
+ * @subnode:	Name of the child node.
+ *
+ * Returns the offset of the child node in DTB.
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/include/kernel/boot.h]: Done

File: core/kernel/boot.c [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+	snprintf(frag, sizeof(frag), "fragment@%d", dt->frag_id);
+	offs = fdt_add_subnode(dt->blob, ioffs, frag);
+	if (offs < 0)
+		return offs;
+
+	dt->frag_id += 1;
+
+	ret = fdt_setprop_string(dt->blob, offs, "target-path", "/");
+	if (ret < 0)
+		return -1;
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/kernel/boot.c]: Fixed in https://github.com/OP-TEE/optee_os/pull/6150/commits/1e00eb9f94a924a9ab913f000eab169f4bcc27c8

File: core/kernel/boot.c [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+}
+#endif /*CFG_EMBED_DTB*/
+
+#if defined(CFG_DT)
+#ifdef _CFG_USE_DTB_OVERLAY
+static int add_dt_overlay_fragment(struct dt_descriptor *dt, int ioffs)
+{
+	char frag[32];
+	int offs;
+	int ret;
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/kernel/boot.c]: Fixed in https://github.com/OP-TEE/optee_os/pull/6150/commits/1e00eb9f94a924a9ab913f000eab169f4bcc27c8

File: core/kernel/dt.c [hunk 10]
```diff
@@ -446,9 +446,9 @@ int dt_getprop_as_number(const void *fdt, int nodeoffset, const char *name,
 #ifdef _CFG_USE_DTB_OVERLAY
 static int add_dt_overlay_fragment(struct dt_descriptor *dt, int ioffs)
 {
-	char frag[32];
-	int offs;
-	int ret;
+	char frag[32] = {0};
```
Comment (etienne-lms) [inline core/kernel/dt.c:449]: replace `{0}` with `{ }`  as per [coding standards](https://optee.readthedocs.io/en/latest/general/coding_standards.html) .

File: core/kernel/dt.c [hunk 11]
```diff
@@ -437,3 +442,239 @@ int dt_getprop_as_number(const void *fdt, int nodeoffset, const char *name,
...
+
+int add_res_mem_dt_node(struct dt_descriptor *dt, const char *name,
+			paddr_t pa, size_t size)
+{
+	int offs = 0;
+	int ret = 0;
+	int addr_size = -1;
+	int len_size = -1;
+	bool found = true;
+	char subnode_name[80] = { 0 };
```
Comment (etienne-lms) [inline core/kernel/dt.c:622]: prefer `{ }`

File: core/kernel/dt.c [hunk 12]
```diff
@@ -437,3 +442,239 @@ int dt_getprop_as_number(const void *fdt, int nodeoffset, const char *name,
...
+boot_final(release_external_dt);
+
+int add_dt_path_subnode(struct dt_descriptor *dt, const char *path,
+			const char *subnode)
+{
+	int offs = 0;
+
+	offs = fdt_path_offset(dt->blob, path);
+	if (offs < 0)
+		return -1;
```
Comment (etienne-lms) [inline core/kernel/dt.c:591]: could return `off` which contains a libftd error number

File: core/kernel/dt.c [hunk 13]
```diff
@@ -437,3 +442,239 @@ int dt_getprop_as_number(const void *fdt, int nodeoffset, const char *name,
...
+		offs = 0;
+	}
+
+	if (IS_ENABLED2(_CFG_USE_DTB_OVERLAY)) {
+		len_size = sizeof(paddr_t) / sizeof(uint32_t);
+		addr_size = sizeof(paddr_t) / sizeof(uint32_t);
+	} else {
+		len_size = fdt_size_cells(dt->blob, offs);
+		if (len_size < 0)
+			return -1;
```
Comment (etienne-lms) [inline core/kernel/dt.c:637]: could return `len_size` that contains a libfdt error number.
All other -1 could be replaced to return a libfdt error number.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: riscv: Get external device tree provided by early boot stage" and
"core: riscv: Add external DT initialization and updating" (with the 2 comments addressed).

Forc the 2 first commits in the series, I'm nitpicking on some functions return value. Original functions implementation is quite old hence these -1. I think it would be better that these functions return meaningfull return codes, either the available libfdt error numbers (these all related to DT operations) or move to `TEE_Result` which is OP-TEE OS  native return code type.
But maybe @jforissier or @jenswi-linaro prefer we preserve the original implementation.

> But some functions such as get_dt() and get_secure_dt() should be independent with CFG_DT.

You can implement `get_dt()`, `get_secure_dt()` and `get_embedded_dt()` in _kernel/dt.c_ and implement their stubs in _kernel/dt.h_ for when `CFG_DT` is disabled.

**Comment by gagachang** [comment]
Hello @etienne-lms 

I submit new commits to move all functions to `core/kernel/dt.c`
If it is OK to you I can rebase it and fix the return values by more commits.

**Review by etienne-lms — COMMENTED** [review]
Ok, please do.

File: core/kernel/dt.c (hunk 11 above)
Comment (gagachang) [reply core/kernel/dt.c]: solved

File: core/kernel/dt.c (hunk 12 above)
Comment (gagachang) [reply core/kernel/dt.c]: solved

File: core/kernel/dt.c (hunk 13 above)
Comment (gagachang) [reply core/kernel/dt.c]: solved

File: core/kernel/dt.c [hunk 14]
```diff
@@ -437,3 +442,284 @@ int dt_getprop_as_number(const void *fdt, int nodeoffset, const char *name,
...
+	}
+
+	ret = snprintf(subnode_name, sizeof(subnode_name),
+		       "%s@%" PRIxPA, name, pa);
+	if (ret < 0 || ret >= (int)sizeof(subnode_name))
+		DMSG("truncated node \"%s@%" PRIxPA"\"", name, pa);
+	offs = fdt_add_subnode(dt->blob, offs, subnode_name);
+	if (offs >= 0) {
+		uint32_t data[FDT_MAX_NCELLS * 2];
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/kernel/dt.c]: solved

**Comment by gagachang** [comment]
Hi @etienne-lms 

Commit https://github.com/OP-TEE/optee_os/pull/6150/commits/31cbd876bd50fc370605c8ceb16c28fa4fa876ee addresses the review.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` cor commit
"core: arm: Move some DT functions to common kernel".

For commit "core: kernel: Refine variable declarations and return values in dt.c". Could you update the inline description of `add_dt_path_subnode()` and `add_res_mem_dt_node()` to state these function return a negative libfdt error number?

**Comment by gagachang** [comment]
Hi @etienne-lms 

Please check https://github.com/OP-TEE/optee_os/pull/6150/commits/656daa26bfa001326c3661559eab31e885d52907#diff-0ed0797a7059fd8959afe054481b76ae89008bf9dbfea5683345846a73879aaaL287

**Review by etienne-lms — COMMENTED** [review]
Thanks for the update.
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: kernel: Refine variable declarations and return values in dt.c".

**Comment by gagachang** [comment]
Thank you @etienne-lms !
Tags are applied

**Comment by gagachang** [comment]
Hello @jforissier 

If this PR is fine, I think it is ready to be merged, thanks!

**Pull request merged**